services/api/src/models/audit-log.js

```diff
@@ -0,0 +1,75 @@
+const mongoose = require('mongoose');
+const { createSchema } = require('../utils/schema');
+const { get, pick, isEmpty, intersection } = require('lodash');
+
+const definition = require('./definitions/audit-log.json');
+
+const schema = createSchema(definition.attributes);
+
+schema.statics.getContextFields = function (ctx) {
+  return {
+    user: ctx.state.authUser?.id,
+    requestMethod: ctx.request.method,
+    requestUrl: ctx.request.url,
+    routeNormalizedPath: ctx.routerPath,
+    routePrefix: ctx.router.opts.prefix,
+  };
+};
+
+schema.statics.getObjectFields = function getObjectFields(object, fields = [], isNew = false) {
+  const isMongooseDoc = object instanceof mongoose.Model;
+  if (!isMongooseDoc) throw Error('AuditLog.getObjectFields only works with mongoose documents');
+
+  const objectFields = {
+    objectId: object.id,
+    objectType: object.constructor.modelName,
+  };
+
+  if (fields.length) {
+    if (isNew) {
+      objectFields.objectAfter = pick(object.toObject({ depopulate: true }), fields);
+    } else {
+      const { auditOriginal, auditPathsModified } = object.$locals;
+      // the mongoose's Document.directModifiedPaths() returns falsely the objectId modified (quite possible due to autopopulate)
+      // this corrects any issues that happens when comparing with objectIds
+      const pathsModified = intersection(auditPathsModified, fields).filter((field) => {
+        if (!object.get(field).equals) return true;
+        return !object.get(field).equals(get(auditOriginal, field));
+      });
+
+      const after = pick(pick(object.toObject({ depopulate: true }), pathsModified), fields);
+
+      if (!isEmpty(after)) {
+        const before = pick(pick(auditOriginal, pathsModified), fields);
+        objectFields.objectAfter = after;
+        objectFields.objectBefore = before;
+      }
+    }
+  }
+
+  return objectFields;
+};
+
+schema.statics.append = function (activity, ctx, { objectId, objectType, objectBefore, objectAfter, type, user }) {
```

Comment:
For input this complex I would rather just have a single options object. Making `ctx` a named option would also make it more explicit what it is... so related to what I was saying above maybe could we do something morel like:

```
function({ object, ctx, ...attributes }) {
  if (object) {
    Object.assign(attributes, this.getObjectFields(object));
  }
  if (ctx) {
    Object.assign(attributes, this.getContextFields(ctx);
  }
  return this.create(attributes);
}
```

Reply:
Also for naming I feel like I would rather it just be `AuditEntry` (to me the entire "file" is the log, not a single line), and I would kind of rather just do `.create` although I see why you did this...maybe it's difficult or not good to shadow the mongoose method... what do you think?

Reply:
I took out the `ctx` of the options. Because i realized `ctx` its required. 
So the split is simpler now, everything in options `AuditLog.append(activity, ctx, options)` is optional

Re the auditEntry sure we could call it that, but i stayed with auditLog because plural form is auditLogs. But yeah its just a name what ever.

Reply:
But `ctx` should not be required no?

Reply:
it should be as it provides the route information

Reply:
but you can have an audit trail outside a route... I thought that was part of the requirements

Reply:
```
For each audit trail entry we want to track:
Originating Client IP address
The user ID that was authenticated at the time
The geographical location (bonus for later)
The API call that the mutation happened on
In case of external integrations the API credentials that were used (future)
```

https://bedrockio.slack.com/archives/C013YV1TPLZ/p1618839594326700?thread_ts=1618824211.322200&cid=C013YV1TPLZ